compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELinalgFunctionOutlining.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/AMDAIEUtils.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "mlir/Conversion/FuncToLLVM/ConvertFuncToLLVM.h"
+#include "mlir/Conversion/FuncToLLVM/ConvertFuncToLLVMPass.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Hoisting.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+
+#define DEBUG_TYPE "iree-amdaie-linalg-function-outlining"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+unsigned uniqueOutlinedMatmul = 0;
+unsigned uniqueOutlinedElementwise = 0;
+
+/// Utility to outline the linalg compute op.
+static FailureOr<func::FuncOp> outlinedToAFunction(
+    IRRewriter &rewriter, ModuleOp moduleOp, linalg::LinalgOp computeOp,
+    std::string outlineFuncName,
+    DenseMap<Operation *, std::string> &computeOpToOutlinedFuncMap) {
+  // Check if the compute op is equivalent to a previously outlined compute op.
```

Comment:
At this point we expect moduleOp's symbol table to not contain `outlineFuncName`. I think this is worth asserting, to avoid to unlikely case that there is already a function with the name you've just created. I suppose you could increment the  uniqueOutlined index in a while loop until an unused name is found.